lib/adyen/hpp/signature.rb

```diff
@@ -0,0 +1,71 @@
+require 'openssl'
+require 'base64'
+
+module Adyen
+  module HPP
+    # The Signature class can sign and verify HMAC signatures for the Hosted Payment Pages
+    class Signature
+      VALID_ALGORITHMS  = %w(sha1 sha256).freeze
+      DEFAULT_ALGORITHM = 'sha256'.freeze
+
+      # Creates a Signature
+      # @param [String] digest_algorithm (optional), defaults to Signature::DEFAULT_ALGORITHM.
+      # Must be one of Signature::VALID_ALGORITHMS
+      # @return [Signature]
+      def initialize(digest_algorithm = nil)
+        digest_algorithm ||= DEFAULT_ALGORITHM
+        raise ArgumentError, "invalid signature digest_algorithm" unless VALID_ALGORITHMS.include?(digest_algorithm)
+        @digest = OpenSSL::Digest.new(digest_algorithm)
+      end
+
+      # Sign the parameters with the given shared secret
+      # @param [Hash] params The set of parameters to sign
+      # @param [String] shared_secret The shared secret for signing/verification
+      # @return [Hash] params The params that were passed in plus a new 'merchantSig' param
+      def sign(params, shared_secret)
+        sig = OpenSSL::HMAC.digest(@digest, Array(shared_secret).pack("H*"), string_to_sign(params))
+        params.merge('merchantSig' => Base64.encode64(sig).strip)
+      end
+
+      # Verify the parameters with the given shared secret
+      # @param [Hash] params The set of parameters to verify. Must include a `merchantSig`
+      #   param that will be compared to the secret calculated here.
+      # @param [String] shared_secret The shared secret for signing/verification
+      # @return [Boolean] true if the `merchantSig` in the params matches our calcuated signature
+      def verify(params, shared_secret)
+        their_sig = params.delete('merchantSig')
+        raise ArgumentError, "params must include 'merchantSig' for verification" if their_sig.empty?
+        our_sig = sign(params, shared_secret)['merchantSig']
+        secure_compare(their_sig, our_sig)
+      end
+
+      private
+
+      def string_to_sign(params)
+        (sorted_keys(params) + sorted_values(params)).map{ |el| escape_value(el) }.join(':')
+      end
+
+      def sorted_keys(hash)
+        hash.sort.map{ |el| el[0] }
+      end
+
+      def sorted_values(hash)
+        hash.sort.map{ |el| el[1] }
+      end
+
+      def escape_value(value)
+        value.gsub(':', '\\:').gsub('\\', '\\\\')
+      end
+
+      def secure_compare(a, b)
+        return false unless a.bytesize == b.bytesize
+
+        l = a.unpack "C#{a.bytesize}"
+
+        res = 0
+        b.each_byte { |byte| res |= byte ^ l.shift }
+        res == 0
+      end
```

Comment:
Why are you using this complicated comparison algorithm? Could `a.force_encoding('binary') == b.force_encoding('binary')` work if you are worried about encodings?